core/DataAccess/LogAggregator.php

```diff
@@ -1083,6 +1083,139 @@ public function queryConversionsByDimension($dimensions = array(), $where = fals
         return $this->getDb()->query($query['sql'], $query['bind']);
     }
 
+    /**
+     * Similar to queryConversionsByDimension and will return data in the same format, but takes into account pageviews
+     * leading up to a conversion, not just the final page that triggered the conversion
+     *
+     * @param string $linkField
+     * @param int $rankingQueryLimit
+     *
+     * @return \Zend_Db_Statement|array
+     */
+    public function queryConversionsByPageView(string $linkField, $rankingQueryLimit = 0)
+    {
+        $dbSettings = new \Piwik\Db\Settings();
+        $tablePrefix = $dbSettings->getTablePrefix();
+        $subQuery = sprintf("
+            MAX((SELECT COUNT(am.idaction)
+             FROM %slog_conversion cam
+             LEFT JOIN %slog_link_visit_action vam ON vam.idvisit = cam.idvisit
+             LEFT JOIN %slog_action am ON am.idaction = vam.idaction_url
+             WHERE cam.idgoal = log_conversion.idgoal AND vam.idvisit = log_link_visit_action.idvisit
+             AND vam.idaction_url IS NOT NULL AND am.type = 1
+             AND vam.server_time <= log_conversion.server_time
+             GROUP BY cam.idgoal, cam.idvisit)) AS `9`
+        ", $tablePrefix, $tablePrefix, $tablePrefix);
```

Comment:
using numbered placeholders makes it possible to use the same placeholder multiple times. Also it might be easier to understand when using constants instead of those "magic" numbers

```suggestion
             FROM %1$slog_conversion cam
             LEFT JOIN %1$slog_link_visit_action vam ON vam.idvisit = cam.idvisit
             LEFT JOIN %1$slog_action am ON am.idaction = vam.idaction_url
             WHERE cam.idgoal = log_conversion.idgoal AND vam.idvisit = log_link_visit_action.idvisit
             AND vam.idaction_url IS NOT NULL AND am.type = %2$s
             AND vam.server_time <= log_conversion.server_time
             GROUP BY cam.idgoal, cam.idvisit)) AS `%3$s`
        ", $tablePrefix, Action::TYPE_PAGE_URL, Metrics::INDEX_GOAL_NB_PAGES_UNIQ_BEFORE);
```
Note: this change might require additional use statements.

Reply:
Definitely easy to understand with constants. :+1: